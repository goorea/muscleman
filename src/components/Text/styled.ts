import styled from 'styled-components/native';

import { fonts } from '@src/theme';

import { TextProps } from './index';

export const StyledText = styled.Text<Omit<TextProps, 'style'>>`
  font-family: ${({ weight }) => fonts[weight || 'normal']};
  font-size: ${({ size }) => `${size || 16}px`};
  color: ${({ theme, color }) => theme[color || 'foreground']};
`;
