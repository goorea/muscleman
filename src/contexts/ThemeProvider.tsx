import React from 'react';
import theme from '@src/theme';
import { useColorScheme } from 'react-native';
import { ThemeProvider as StyledThemeProvider } from 'styled-components/native';
import { ThemeContextState } from '@src/types/theme';

export const ThemeContext = React.createContext<ThemeContextState>({
  dark: false,
  toggleTheme: () => null,
  colors: theme.light,
});

const ThemeProvider: React.FC = ({ children }) => {
  const dark = useColorScheme() === 'dark';
  const [themeState, setTheme] = React.useState<ThemeContextState>({
    dark,
    toggleTheme: () => {
      setTheme(prevState => ({
        dark: !prevState.dark,
        toggleTheme: prevState.toggleTheme,
        colors: prevState.dark ? theme.light : theme.dark,
      }));
    },
    colors: dark ? theme.dark : theme.light,
  });

  React.useEffect(() => {
    if (dark !== themeState.dark) {
      themeState.toggleTheme();
    }
  }, [dark, themeState]);

  return (
    <ThemeContext.Provider value={themeState}>
      <StyledThemeProvider theme={themeState.colors}>
        {children}
      </StyledThemeProvider>
    </ThemeContext.Provider>
  );
};

export function useTheme(): ThemeContextState {
  return React.useContext<ThemeContextState>(ThemeContext);
}

export default ThemeProvider;
