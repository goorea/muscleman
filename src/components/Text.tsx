import React from 'react';
import styled from 'styled-components/native';
import { TextStyle } from 'react-native';
import { fonts } from '@src/theme';
import { Colors } from '@src/types/theme';

export type TextProps = {
  weight?: 'normal' | 'bold' | 'thin';
  size?: number;
  color?: keyof Colors;
  style?: TextStyle;
  ellipsis?: boolean;
};

const RNText = styled.Text<Omit<TextProps, 'style'>>`
  font-family: ${({ weight }) => fonts[weight || 'normal']};
  font-size: ${({ size }) => `${size || 16}px`};
  color: ${({ theme, color }) => theme[color || 'foreground']};
`;

const Text: React.FC<TextProps> = ({
  weight,
  size,
  color,
  style,
  ellipsis = false,
  children,
}) => {
  return (
    <RNText
      weight={weight}
      size={size}
      color={color}
      style={style}
      numberOfLines={ellipsis ? 1 : undefined}>
      {children}
    </RNText>
  );
};

export default Text;
