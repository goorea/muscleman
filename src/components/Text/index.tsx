import { isEqual } from 'lodash';
import React from 'react';
import { TextStyle } from 'react-native';

import { Colors } from '@src/types/theme';

import { StyledText } from './styled';

export type TextProps = {
  children: React.ReactNode;
  weight?: 'normal' | 'bold' | 'thin';
  size?: number;
  color?: keyof Colors;
  style?: TextStyle;
  ellipsis?: boolean;
};

const Text: React.FC<TextProps> = ({
  weight,
  size,
  color,
  style,
  ellipsis = false,
  children,
}) => (
  <StyledText
    weight={weight}
    size={size}
    color={color}
    style={style}
    numberOfLines={ellipsis ? 1 : undefined}>
    {children}
  </StyledText>
);

export default React.memo<TextProps>(Text, isEqual);
