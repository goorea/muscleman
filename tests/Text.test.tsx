import 'react-native';
import React from 'react';
import { render } from '@testing-library/react-native';
import Text, { TextProps } from '@src/components/Text';
import ThemeProvider from '@src/contexts/ThemeProvider';

describe('Text 컴포넌트', () => {
  const rendered = (props?: TextProps) =>
    render(
      <ThemeProvider>
        <Text {...props} />
      </ThemeProvider>,
    );

  it('렌더링이 올바르게 된다', () => {
    const { toJSON } = rendered();

    expect(toJSON()).toMatchSnapshot();
  });

  it('weight, size, color, style, ellipsis props를 가지고 있다', () => {
    const { container } = rendered({
      weight: 'bold',
      size: 20,
      color: 'primary',
      style: {
        marginTop: 10,
      },
      ellipsis: true,
    });

    expect(container.props.children.props).toHaveProperty('weight');
    expect(container.props.children.props).toHaveProperty('size');
    expect(container.props.children.props).toHaveProperty('color');
    expect(container.props.children.props).toHaveProperty('style');
    expect(container.props.children.props).toHaveProperty('ellipsis');
  });
});
