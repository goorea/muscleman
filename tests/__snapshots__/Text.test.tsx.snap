// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text 컴포넌트 렌더링이 올바르게 된다 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#000000",
        "fontFamily": "SpoqaHanSansNeo-Regular",
        "fontSize": 16,
      },
    ]
  }
/>
`;
